qutip/core/cy/coefficient.pyx

```diff
@@ -24,11 +24,14 @@ cdef class Coefficient:
     cpdef void arguments(self, dict args) except *:
         self.args.update(args)
 
-    def __call__(self, double t, dict args={}):
-        """update args and return the
+    def __call__(self, double t, dict args=None):
+        """Update args and return the coefficient value at `t`.
         """
+        cdef Coefficient cp
         if args:
-            self.arguments(args)
+            cp = self.copy()
+            cp.arguments(args)
+            return cp._call(t)
```

Comment:
It's odd that we have to create a new `Coefficient` to run a method on this instance. The question is what is `self.args`? Is it the fixed arguments for this Coefficient object? Or is it the default arguments only? And what are the args to `__call__`? Are they all the arguments, or just additional ones?

One approach would be to change `_call` to take `args` as a paramter and do `all_args = self.args.copy(); all_args.update(args)` (with better variable names, of course). Or to pass the full args to `_call` and do the argument updating here.

Thoughts?

Reply:
As I see it, there are two problems at play - one of call speed (`Coefficient`'s primary purpose is C-level fast calling) and one of mutability.  We attempt to avoid reads to an `args` dictionary within each call, but instead read it once at the most recent time they're set, and bind them to named variables within the C struct (created by our codegen).  Then call speed is faster from them, since the typical use-case will involve many many calls to the `Coefficient` via something like `mesolve` before any `args` are changed.

Directly calling `Coefficient` while passing `args` is not going to happen very much at all, but still ought to be supported just for logical consistency within the library - there are various points at which you can define the `args` in functions that indirectly call `Coefficient`, so it's only natural that direct calls should work too, and much easier for the user if we handle the state-dependent boilerplate.

That said, I mentioned in a different comment that the current implementation really has mutating-state problems (see https://github.com/qutip/qutip/pull/1481#discussion_r609744805).  This `copy` means that we're expending more GC cycles to collect stuff (and hence we shouldn't be using this form in inner loops, but we already don't), but it does maintain the expected scoping rules.  If you have another reference to this same coefficient, _this_ way won't allow a single apparently-pure function call to have side effects.

Reply:
If you have a `Coefficient`, say `sin(w*t)`, it store it's value of `w` so it can be called with `coeff(t)`.
But you can call it with different value of `w` by passing it in an `args` dictionary.
This only applied to that single call and leave the object `args` untouched.

I feel this feature will be mostly used in tests and maybe some debugging by users trying to see if their `QobjEvo` are behaving as expected...
So speed should not matter much here and making a copy is the simplest way to ensure it does not to have any side effect.
If we have a need for efficiency here, we can move the logic in `_call`.

Reply:
Maybe then a better interface would be the following:
```python
def replace(self, args=None):
    cp = self.copy()
    if args:
        cp.arguments(args)
    return cp

def __call__(self, t):  # note lack of args
      ...
```
Then users can either call `.arguments(...)` if they need to update the arguments, or `.replace(...)` if they need a new Coefficient instance with update arguments and `__call__` is always fast because the arguments have already been updated before.

Thoughts? Also not necessary for this to happen in this PR perhaps -- it can just be an idea for the future.

Reply:
P.S. `.replace(...)` is inspired by `datetime.datetime.replace`.

Reply:
I like it, and personally I'd go all the way and have `Coefficient` be a completely immutable type. I'd remove `Coefficient.arguments` and only have the suggested `Coefficient.replace` - I'm always really put off by having to deal with mutability in any form, because you can almost always end up with funny cases where stuff can mutate from underneath you (like I demonstrated in that comment I linked above).  In stuff I write for myself, I keep everything immutable unless it _absolutely_ has to be mutable for performance.  Maybe it's just because I prefer it, but I do find that immutability-by-default eliminates a whole class of side-effect-dependent bugs that OOP often struggles with.

I think that the way we now manage compiled Cython coefficients means that the copy operation should be very fast, and this is never going to be inner-loop code anyway, so there's likely little-to-no measurable performance gain from allowing the mutation if we do it right.

Reply:
A few things I see:
- There is two situation that `Coefficient.arguments` could be in a loop: feedback and control.
For feedback, the user should expect an impact on speed: time to create a `Qobj` for the state is similar to copying a `Coefficient`.
For control, they want to be able to step through the evolution changing the `args` at each timestep.
- Also array based `Coefficient` will be the slowest to `replace` with the copy the arrays but have not `args` to update.
- Copying string array is slow presently: it uses `pickle.loads(pickle.dumps(self))`
- I am not the best with names but `replace` feels wrong. What are we replacing?

Looking at timing, with a proper `copy` for string coefficient and `replace` returning a copy only for mutable ones (`return self` for array based ones), this should be mostly unnoticeable even in evolution with feedback. (If we see a performance issue there later, we could always remake `arguments` as a `cdef` method.)

Reply:
Array coefficients won't need to copy the arrays to replace the arguments; as long as we maintain that the arrays are never mutated after the coefficient is created, the only references to them will be held inside `Coefficient`, so there's no problem there.  If copying of string arrays is slow enough to be bottleneck, we can look into speeding it up - the internal C callable has already been Cythonised and compiled, so at its heart the only work that needs to be done is to allocate a new Python object and copy a few references.

If we move to a model where `Coefficient` is immutable, then `replace` is replacing one or more scope variables - it's got similarities with the `_replace` method on named tuples:
```python
import collections
Record = collections.namedtuple('Record', ['name', 'age'])
jake = Record('jake', 26)
twin = jake._replace(name='john')
```
and a handful of other standard library types.  I always preferred `with`, personally, but we can't use that because it's a keyword :(

Reply:
I am sold on making `Coefficient` immutable, and `QobjEvo`'s `_BaseElement` with it. 
The copying of string coefficient is 50x the copying of function coefficient, so it is worth to rework it, but we need to rework the codegen anyway to remove the arguments function.

With the logic of `replace`, it feels like it should do more that just act on the `args` and be able to update the `tlist` or the like...

Reply:
Oh yeah, good idea - it can be very like the `namedtuple` one then.  The signature becomes
```python
def replace(self, *, arguments=None, tlist=None, ...):
    ...
```
so the calls look like
```python
coeff = qutip.coefficient(...)
coeff.replace(arguments={...})
```

Reply:
This all got implemented in #1507 -- thanks!